pkg/trace/writer/trace.go

```diff
@@ -172,27 +175,33 @@ func (w *TraceWriter) flush() {
 		log.Errorf("Failed to serialize payload, data dropped: %v", err)
 		return
 	}
-	p := newPayload(map[string]string{
-		"Content-Type":     "application/x-protobuf",
-		"Content-Encoding": "gzip",
-		headerLanguages:    strings.Join(info.Languages(), "|"),
-	})
-	gzipw, err := gzip.NewWriterLevel(p.body, gzip.BestSpeed)
-	if err != nil {
-		// it will never happen, unless an invalid compression is chosen;
-		// we know gzip.BestSpeed is valid.
-		log.Errorf("gzip.NewWriterLevel: %d", err)
-		return
-	}
-	gzipw.Write(b)
-	gzipw.Close()
 
 	atomic.AddInt64(&w.stats.BytesUncompressed, int64(len(b)))
 	atomic.AddInt64(&w.stats.BytesEstimated, int64(w.bufferedSize))
 
-	for _, sender := range w.senders {
-		sender.Push(p)
-	}
+	w.wg.Add(1)
+	go func() {
```

Comment:
I'm not too keen on having this unbounded. What would prevent you from possibly accumulating hundreds of megabytes of data in parallel compressions (say, this writer has to process traces with lots of metadata, each of 3MB and so each would trigger a separate flush)?

This probably won't happen on fast underloaded machines since the compression should be fast enough but on machines with very high load I could imagine this happening.

I think the savings you've observed in terms of allocations might be a symptom of some other queues being too large (and thus the reduction of blocking time leading to less filling of said queues?).

Reply:
> I think the savings you've observed in terms of allocations might be a symptom of some other queues being too large (and thus the reduction of blocking time leading to less filling of said queues?).

The savings in terms of memory are a reflection of the number of the unbounded [sampling goroutine](https://github.com/DataDog/datadog-agent/blob/dd92d1a485479428dff146df76b0188e4134cdd4/pkg/trace/agent/agent.go#L279) which grows indefinitely in memory when the writer is blocking. The writer should never block. If it ends up blocking again (due to limited goroutines or filled buffered channels), the memory will start bubbling up elsewhere in the pipeline, possibly even in the client app. The alternative is to drop, which is also bad.

> I'm not too keen on having this unbounded. What would prevent you from possibly accumulating hundreds of megabytes of data in parallel compressions (say, this writer has to process traces with lots of metadata, each of 3MB and so each would trigger a separate flush)?

Me neither, but I think it's the best choice for these reasons:
* We can't assume how many of these we need. Each customer has different traffic.
* We can't prove that having them unbounded is a problem. e.g.: previous 54k spans/s was nearly impossible, now it is. That's a huge win isn't?
* We already have unbounded goroutines everywhere in the trace-agent (processing, normalizing, sender). They have not yet presented to be the bottlenecks in any tests I ran. We will likely (maybe) get there.
* I've experimented with a [dynamically growing pool](https://github.com/DataDog/datadog-agent/blob/gbbr/workers/pkg/trace/agent/work/work.go) but it didn't present any benefits.
* We have the watchdog system which rate limits API intake based on how much memory the trace agent is using.

I think there is no need to have these bounded to a magical number. It could be dynamic, using the pool I linked to, but it hasn't proven to provide noticeable benefits.

Reply:
Fair points